src/tools/chip-cert/CertUtils.cpp

```diff
@@ -773,39 +780,105 @@ bool MakeAttCert(AttCertType attCertType, const char * subjectCN, uint16_t subje
         ReportOpenSSLErrorAndExit("X509_set_pubkey", res = false);
     }
 
-    // Add common name attribute to the certificate subject DN.
-    if (!X509_NAME_add_entry_by_NID(X509_get_subject_name(newCert), NID_commonName, MBSTRING_UTF8,
-                                    reinterpret_cast<unsigned char *>(const_cast<char *>(subjectCN)),
-                                    static_cast<int>(strlen(subjectCN)), -1, 0))
+    // Encode Common Name (CN) Attribute.
     {
-        ReportOpenSSLErrorAndExit("X509_NAME_add_entry_by_NID", res = false);
-    }
+        char cnAttrStr[chip::Crypto::kMax_CommonNameAttr_Length];
+        size_t cnAttrStrLen = 0;
+        if (subjectCN != nullptr)
+        {
+            VerifyOrReturnError(strlen(subjectCN) <= sizeof(cnAttrStr), false);
+            memcpy(cnAttrStr, subjectCN, strlen(subjectCN));
+            cnAttrStrLen += strlen(subjectCN);
+        }
 
-    // Add VID attribute to the certificate subject DN.
-    if (subjectVID != VendorId::NotSpecified)
-    {
-        char chipAttrStr[5];
+        if (encodeVIDandPIDasCN)
+        {
+            if (subjectVID != VendorId::NotSpecified)
+            {
+                // Add space to separate from the previous string.
+                if (cnAttrStrLen > 0)
+                {
+                    VerifyOrReturnError((cnAttrStrLen + 1) <= sizeof(cnAttrStr), false);
+                    cnAttrStr[cnAttrStrLen] = ' ';
+                    cnAttrStrLen++;
+                }
+
+                VerifyOrReturnError((cnAttrStrLen + strlen(chip::Crypto::kVIDPrefixForCNEncoding) +
+                                     chip::Crypto::kVIDandPIDHexLength) <= sizeof(cnAttrStr),
+                                    false);
+
+                memcpy(&cnAttrStr[cnAttrStrLen], chip::Crypto::kVIDPrefixForCNEncoding,
+                       strlen(chip::Crypto::kVIDPrefixForCNEncoding));
```

Comment:
This code has a lot of `memcpy` and `strlen` arithmetic and is error-prone. Since this is a tool, *I strongly suggest using `std::string` instead*.

Reply:
Will do as a followup.